src/serialize.rs

```diff
@@ -47,125 +49,153 @@ impl MessageReader for OwnedSpaceMessageReader {
     }
 }
 
-pub fn new_reader<U : InputStream>(
-    input_stream : &mut U,
-    options : ReaderOptions) -> Result<OwnedSpaceMessageReader> {
-
-    let mut first_word : [WireValue<u32>; 2] = unsafe {::std::mem::uninitialized()};
-    unsafe {
-        let ptr : *mut u8 = ::std::mem::transmute(first_word.as_mut_ptr());
-        let buf = ::std::slice::from_raw_parts_mut::<u8>(ptr, 8);
-        try!(input_stream.read_exact(buf));
-    }
+/// Read a Cap'n Proto serialized message from a stream with the provided options.
+///
+/// For optimal performance, `read` should be a buffered reader type.
+pub fn read_message<R>(read: &mut R, options: ReaderOptions) -> Result<OwnedSpaceMessageReader>
+where R: Read {
+    let (total_words, segment_slices) = try!(read_frame(read, options));
+    read_segments(read, total_words, segment_slices, options)
+}
 
-    let segment_count : u32 = first_word[0].get() + 1;
+/// Reads a stream frame from `read` and returns the total number of words across all
+/// segments, as well as the segment offsets.
```

Comment:
I think using the word "frame" like this is confusing. To me, a "frame" includes the payload as well as the header. The standard Cap'n Proto jargon for what this function reads is the "segment table", though I suppose that "frame header" would be an acceptable name too.